File: lib/services/person_repository.dart
import 'dart:convert';
import 'dart:io';

import 'package:path_provider/path_provider.dart';

import '../models/person.dart';

/// 人物身份的本地持久化仓库（JSON 文件 + 头像图片目录）。
class PersonRepository {
  static const String _fileName = 'people.json';

  final List<Person> _people = [];
  bool _loaded = false;
  Directory? _docsDir;

  List<Person> get people => List.unmodifiable(_people);

  Future<Directory> _dir() async {
    return _docsDir ??= await getApplicationDocumentsDirectory();
  }

  Future<File> _file() async {
    final dir = await _dir();
    return File('${dir.path}/$_fileName');
  }

  /// 头像保存目录。
  Future<Directory> avatarsDir() async {
    final dir = await _dir();
    final avatars = Directory('${dir.path}/avatars');
    if (!await avatars.exists()) {
      await avatars.create(recursive: true);
    }
    return avatars;
  }

  Future<void> load() async {
    if (_loaded) return;
    try {
      final file = await _file();
      if (await file.exists()) {
        final raw = await file.readAsString();
        final list = jsonDecode(raw) as List<dynamic>;
        _people
          ..clear()
          ..addAll(list
              .map((e) => Person.fromJson(Map<String, dynamic>.from(e as Map))));
      }
    } catch (_) {
      // 读取失败时以空库启动，避免阻塞应用。
    }
    _loaded = true;
  }

  Future<void> _save() async {
    final file = await _file();
    final data = jsonEncode(_people.map((p) => p.toJson()).toList());
    await file.writeAsString(data);
  }

  Future<void> upsert(Person person) async {
    final idx = _people.indexWhere((p) => p.id == person.id);
    if (idx >= 0) {
      _people[idx] = person;
    } else {
      _people.add(person);
    }
    await _save();
  }

  Future<void> delete(String id) async {
    final person = _people.where((p) => p.id == id).cast<Person?>().firstOrNull;
    _people.removeWhere((p) => p.id == id);
    // 顺带删除头像文件
    final path = person?.avatarPath;
    if (path != null) {
      try {
        final f = File(path);
        if (await f.exists()) await f.delete();
      } catch (_) {}
    }
    await _save();
  }
}

extension _FirstOrNull<E> on Iterable<E> {
  E? get firstOrNull {
    final it = iterator;
    if (it.moveNext()) return it.current;
    return null;
  }
}

File: xbot-native/app/src/main/java/com/xbot/android/store/PersonRepository.kt
package com.xbot.android.store

import android.content.Context
import android.util.Log
import com.xbot.android.model.Person
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import java.io.File

/**
 * 人物身份的本地持久化仓库（JSON 文件 + 头像图片目录）。
 * 对应 Flutter person_repository.dart。
 *
 * - people.json：全部人物（含 embeddings）
 * - avatars/：每人一张头像 jpg（裁剪后的人脸）
 */
class PersonRepository(private val context: Context) {

    companion object {
        private const val TAG = "PersonRepository"
        private const val FILE_NAME = "people.json"
    }

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val peopleList = ArrayList<Person>()
    private var loaded = false

    /** 当前人物库（不可变视图）。 */
    val people: List<Person> get() = peopleList.toList()

    private fun file(): File = File(context.filesDir, FILE_NAME)

    /** 头像保存目录。 */
    fun avatarsDir(): File {
        val dir = File(context.filesDir, "avatars")
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    /** 加载人物库；读取失败以空库启动，不阻塞应用。 */
    fun load() {
        if (loaded) return
        try {
            val f = file()
            if (f.exists()) {
                val raw = f.readText()
                val list = json.decodeFromString(ListSerializer(Person.serializer()), raw)
                peopleList.clear()
                peopleList.addAll(list)
                Log.i(TAG, "已加载 ${peopleList.size} 个人物")
            }
        } catch (e: Exception) {
            Log.e(TAG, "读取 people.json 失败: ${e.message}")
        }
        loaded = true
    }

    private fun save() {
        try {
            file().writeText(json.encodeToString(ListSerializer(Person.serializer()), peopleList))
        } catch (e: Exception) {
            Log.e(TAG, "写入 people.json 失败: ${e.message}")
        }
    }

    fun upsert(person: Person) {
        val idx = peopleList.indexOfFirst { it.id == person.id }
        if (idx >= 0) peopleList[idx] = person else peopleList.add(person)
        save()
    }

    fun delete(id: String) {
        val person = peopleList.firstOrNull { it.id == id }
        peopleList.removeAll { it.id == id }
        // 顺带删除头像文件。
        person?.avatarPath?.let { path ->
            try { File(path).takeIf { it.exists() }?.delete() } catch (_: Exception) {}
        }
        save()
    }

    fun get(id: String): Person? = peopleList.firstOrNull { it.id == id }
}
